Extract bundled files when IncludeAllContentForSelfExtract is set

mateoatr opened the pull request:
Fixes #42352

Commit:
Extract bundled files when IncludeAllContentForSelfExtract is set

agocke commented on src/installer/corehost/cli/bundle/extractor.cpp:
```diff
@@ -204,7 +204,7 @@ void extractor_t::extract_new(reader_t& reader)
     for (const file_entry_t& entry : m_manifest.files) 
     {
-        if (entry.needs_extraction())
+        if (m_manifest.is_netcoreapp3_compat_mode() || entry.needs_extraction())
```
It looks like there are a number of other places in this file that are wrong (`locate`, et al). I think a safer approach than trying to modify all the use sites is to have `file_entry_t::read` take an extra `force_extract` parameter, then pass it through to each `file_entry_t` constructor and store it in a field. Then we can check the field in `needs_extraction` and avoid changing all the call sites

vitek-karas replied:
Agreed - the problem is everywhere we call `entry.needs_extraction`, so it would be easier to bake it into the `needs_extraction` itself. It would feel cleaner if we stored something like `extracted` on each file entry.

agocke commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/NetCoreApp3CompatModeTests.cs:
```diff
@@ -0,0 +1,73 @@
+                .Should()
+                .Pass()
+                .And
+                .HaveStdOutContaining("Hello World");
```
Any way we could also confirm that the assemblies are actually loaded from that location? Perhaps by printing the result of `Assembly.Location` or similar?

elinor-fung replied:
We should definitely test for this.

I think we always set `BUNDLE_PROBE` (regardless of 3.x compat mode), so bundle probing takes precedence during assembly loading - I could see some undesired behaviour happening as a result.

vitek-karas replied:
With the most recent changes bundle probe should return false for all queries - I checked that the runtime behavior in that case is the same as without bundle probe in the case of assembly resolution. Note that it's not the same in other cases - especially Environment.GetCommandLineArguments[0] should still return the .exe path, not the main .dll path.

I also checked all the other places in runtime where we somehow react to bundle probe, and they all look correct to me.

elinor-fung commented on src/installer/corehost/cli/bundle/manifest.h:
```diff
@@ -16,16 +17,28 @@ namespace bundle
     private:
-        bool m_need_extraction;
+        bool m_files_need_extraction;
+        bool m_netcoreapp3_compat_mode;
```
Actually, what exactly is netcoreapp3 compat mode / how far does 'compat' go? The comments for `header_flags_t` call out extraction and `AppContext.BaseDirectory`. We also modified some other behaviours (e.g. `GetCommandLineArgs`, `APP_CONTEXT_DEPS_FILES`) - those aren't part of compat mode?

agocke replied:
My understanding was that those behaviors should be based on whether the containing assembly/file was loaded from the bundle. If it works based on whether the file was merely present in the bundle, that could be a problem.

elinor-fung replied:
The APIs on `Assembly` are based on if the assembly was loaded from the bundle, but `Environment.GetCommandLineArgs` and the `APP_CONTEXT_DEPS_FILES` property are not tied to an assembly. I believe they are just based on whether or not we are running a bundle. (Not sure if there have been other behaviour changes that fall into the same category.)

agocke replied:
Yeah I see what you mean. It's probably worth auditing every call to `is_single_file_bundle`, since that API alone doesn't differentiate between a bundle for extraction, and a bundle for in-memory load.

Commit:
Don't pass bundle probe to the runtime for netcoreapp3 compat mode

vitek-karas committed:
Bundle probe should not report extracted files

This also means we can still pass bundle probe to the runtime even in full-extract case as all files will be marked as extracted in that situation (so the probe should return false all the time).
This might be benefitial as the runtime will still "know" it's single-file (we don't really use that knowledge today though).

Improve the test to actually validate locally built binaries and add validation of TPA and framework assembly loading.

Commit:
CodeBaseThrows should fail

vitek-karas committed:
Normalize app base path

vitek-karas committed:
Merge branch 'netCoreApp3Compat' of https://github.com/mateoatr/runtime into netCoreApp3Compat

vitek-karas commented:
I tested lot of cases around the APIs which behave in a different/interesting way in single-file cases - AFAIK the behavior is as expected now. That is - for 3.1 backcompat, almost all APIs behave as non-single, with a couple of exceptions.

@elinor-fung @agocke  could you please take a second look?

elinor-fung commented on src/installer/corehost/cli/hostpolicy/hostpolicy_context.cpp:
```diff
@@ -166,7 +166,8 @@ int hostpolicy_context_t::initialize(hostpolicy_init_t &hostpolicy_init, const a
-    pal::string_t app_base = resolver.get_app_dir();
+    pal::string_t app_base;
+    resolver.get_app_dir(&app_base);
     coreclr_properties.add(common_property::TrustedPlatformAssemblies, probe_paths.tpa.c_str());
```
At line 129 above, we explicitly add `System.Private.CoreLib` to the TPA for non-single file or if it is not in the bundle. Do we need to keep doing that in compat mode? (I don't actually know why we need to do that in general)

vitek-karas replied:
As far as I can tell the code at line 129 is not needed anymore - but it has been there since forever. So I would say we leave it as is for 5. We can definitely discuss removing it for 6 - mainly because runtime actually doesn't use TPA to look for corelib (or with my latest change it will only do so as the last resort). And as far as I can tell SDK will put corelib into `.deps.json` always anyway.

The other thing is that runtime will use the first occurrence for each assembly simple name, since we're appending here this value is basically never used for anything.

jkotas replied:
> mainly because runtime actually doesn't use TPA to look for corelib

For 6, I would rather change the runtime to always use TPA for corelib and get rid of the other paths.

vitek-karas replied:
I agree - just the bundle check and TPA - but it also needs refactoring to not parse TPA twice...

elinor-fung reviewed (commented):
In `AssemblyBinder::BindToSystem`, if the bundle probe doesn't return a valid location, we expect `System.Private.CoreLib.dll` to be in the runtime's directory - won't that be problematic for platforms like Linux where the runtime is linked into the host?

> almost all APIs behave as non-single, with a couple of exceptions

Can we make sure we have a list somewhere in comments and for doc?

elinor-fung commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/NetCoreApp3CompatModeTests.cs:
```diff
@@ -0,0 +1,91 @@
+                    .PublishProject(runtime: TestFixture.CurrentRid, outputDirectory: BundleHelper.GetPublishPath(TestFixture));
+
+                // The above will publish the app as self-contained using the stock .NET Core SDK (typically from Program Files)
+                // In order to test the hosting components we need the hostpolicy and hostfxr built by the repo
```
Oh, wait, these tests are not using `singlefilehost`? Is that intentional? It seems like - especially since this is specifically for backcompat - we should be using the new single-file host.

vitek-karas replied:
I'm looking into this - it's broken in more than one way unfortunately. The way the tests work was viable in 3.1, but it's not going to work in 5.0 at all.

elinor-fung commented:
Re: testing - not sure if this has already been done or how simple it would be, but it might be interesting to grab a known consumer of 3.x single-file and try building/running it against 5.0 with this backcompat property. I remember Greenshot being a common reporter of issues with single-file in 3.x

vitek-karas commented:
@elinor-fung You're correct that this is broken on Linux due to not finding CoreLib. Unfortunately this is a bigger issue:
In runtime there's a notion of "system directory", which currently is always deduced from the location of the runtime module. For non-single-file this points to the path of the coreclr.dll. For single-file this points to the path of the bundle .exe. This works well for all cases except this special case - since we want the system to behave as if the app is loaded from the extracted location fully.

CoreLib's path is derived from the system path.
There are several other places where system path is used - for example that's where we look for DAC, that's where GC looks for plugin GCs, that's where we look for runtime resources, ....

I can see several possible ways to "fix" this:
* Introduce a fallback when we look for corelib and look into the path passed by APP_CONEXT_BASE_DIRECTORY runtime property - which will point to the extraction directory. This would fix the CoreLib problem, but the other dependencies on system path would still be broken.
* Introduce a new runtime property SYSTEM_PATH - the host would set this only in this special case and we would use that everywhere. This would simulate the old behavior pretty well.
* Change how we package single-file in the 3.1 case - basically go back to 3.1 behavior fully - so don't use superhost - bundle coreclr.dll as normal.
  * We could use the 3.1 apphost for this (the one which can self-extract) - but it would be a servicing nightmare probably
  * We would have to introduce a new binary on linux - superhost can't do this, so we would have to build the singlefilehost just like Windows does on Linux as well - and ship that.

If I could get any solution for free I would pick the last one - go back to exactly how 3.1 did it, but that's very expensive and introduces lot of differences in behavior in 5.0.
The next best solution right now seems to introduce the SYSTEM_PATH property and wire it up internally in the runtime. It's still not ideal, but it's better than hacking around CoreLib alone.

vitek-karas commented:
Yet another possible solution would be to hardcode CoreLib into the bundle such that the probe always returns true for CoreLib (even if it's extracted onto disk) - the downside is that corelib would be memory-loaded so without Location and so on - less compatible.

vitek-karas commented:
Candidate fix for the CoreLib issue on Linux - passing system path from host approach:
https://github.com/mateoatr/runtime/pull/1

It works well on Linux - also works on Windows (since it will sue the same codepath on Windows as well for consistency).
Please take a look - I didn't want to push it here if we don't think it's the right way to fix the problem.

jkotas commented:
> Change how we package single-file in the 3.1 case - basically go back to 3.1 behavior fully - so don't use superhost - bundle coreclr.dll as normal

This is the most compatible solution. It would be my top pick. Note that it will also address all not-yet-known issues where the true single file is breaking layout assumptions hardcoded in components out there (e.g. debugger?).

The next best solution that I can think of is to extra the CoreLib path from `TRUSTED_PLATFORM_ASSEMBLIES`. The host send the CoreLib path in `TRUSTED_PLATFORM_ASSEMBLIES`, but runtime ignores it and tries to find corelib next to coreclr.dll.

elinor-fung commented:
> If I could get any solution for free I would pick the last one - go back to exactly how 3.1 did it

Agreed. But, but yeah - the host build / packaging / sdk changes might be a bit expensive at this point.

The CoreLib from `TRUSTED_PLATFORM_ASSEMBLIES` solution would only hit the CoreLib problem, but not other dependencies on the system path? Or would we be using that as the system path as well?

jkotas commented:
I would fix just the CoreLib problem and leave the rest alone (for the 5.0 fix). The system directory concept is overloaded and it would be best to eliminate it.

Commit:
Keep deps.json in APP_CONTEXT_DEPS_FILES for .NET Core 3.x compat mode

Commit:
Only expect app.deps.json

vitek-karas committed:
Search for CoreLib in TPA

vitek-karas commented:
So I prepared a fix which adds a fallback to CoreLib search - if we don't find it in the bundle, nor "next" to the runtime, we will also look for it in TPA and use that path there.

Tested it that it works in the single-file scenario on Linux.

The change is not super clean as doing that would require non-trivial refactorings, so I avoided that. I just did the minimum to avoid too much code duplication. The general idea:
* Use the Configuration class to get to the TPA value (this is also not very clean, but it works and effectively does the right thing)
* Use the same code we use to parse TPA - we can now end up parsing it twice
* Search of CoreLib in the TPA that way

The change is here (didn't push it here to get some feedback on the approach first)
https://github.com/mateoatr/runtime/pull/2

@elinor-fung @jank could you please take a quick look if this approach sounds reasonable?

jkotas commented:
Looks reasonable as release/5.0 fix.

elinor-fung commented:
That seems reasonable for 5.0

mateoatr committed:
Merge pull request #2 from vitek-karas/CoreLibFromTPA

Search for CoreLib in TPA

vitek-karas commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/NetCoreApp3CompatModeTests.cs:
```diff
@@ -0,0 +1,91 @@
+                    .EnsureRestoredForRid(TestFixture.CurrentRid, RepoDirectories.CorehostPackages)
+                    .PublishProject(runtime: TestFixture.CurrentRid, outputDirectory: BundleHelper.GetPublishPath(TestFixture));
+
+                // The above will publish the app as self-contained using the stock .NET Core SDK (typically from Program Files)
```
As far as I know dotnet/runtime switched to a mode where if the machine-wide has high enough version installed, it will use that and not download its own into .dotnet. The download only happens if the machine-wide doesn't have the right version.

Regardless, the point is actually that we get hostpolicy from non-live build (machine-wide or .dotnet, both are downloaded from official builds, not from the local machine) - and we need to test live-built binaries.

vitek-karas committed:
PR feedback

vitek-karas committed:
Fix bundle tests to actually test single-file bundles

The tests were basically running everything using apphost.exe - the reason the tests worked is basically a weird coincidence (the bundle was always self-contained so it had everything managed in it and all the native dlls were beside the .exe) - so from host's point of view this looked like a self-contained app - and then hostpolicy/hostfxr noticed it's also a bundle and treated it as such.

This change introduces new test baseclass to reduce duplication and standardizes the tests to mostly use self-contained publish then bundled with singlefilehost.exe. Currently they do leave hostpolicy.dll and hostfxr.dll next to the exe (they're written in .deps.json so they must exist on disk) but at runtime they won't be used since the singlefilehost.exe has them bundled in itself.

elinor-fung committed:
Don't use Microsoft.NETCore.App.Internal package

From 3.0+, Microsoft.NETCore.App just chains in the appropriate runtime
packs. The .Internal package was necessary for tests to use when we were
building with a older SDK, but targeting 3.0+. We should be able to get
rid of the workaround now.

This actually causes problems with the full extraction single-file mode
because pre-5.0 and in the .Internal package, System.Private.CoreLib.dll
was not listed as a runtime assembly (only under native). As a result,
it would not end up in the TPA for a single-file bundle.

elinor-fung committed:
AppWithSubDirs: add back RuntimeIdentifier, don't use .Internal package

The test assets are all restored - passing in TestTargetRid - before
the test runs. Since we do a rid-specific publish later, it needs to
be restored with the approriate rid.

Stop using .Internal package. This resulted in host binaries getting
published even for framework-dependent publishes.

vitek-karas committed:
Remove unused method

vitek-karas committed:
Adapt to recent test changes

vitek-karas commented:
I ported changes from https://github.com/dotnet/runtime/pull/42611 into this PR now - fixes presence of app.deps.json in the runtime property (Since in the compat mode the .deps.json file actually exists on disk, so we can return path to it).

vitek-karas committed:
Rework the single-file API tests for better coverage

The tests now validate behavior of all of the interesting APIs in both single-file and backcompat mode.
Added the AppContext.BaseDirectory to the tests.

vitek-karas commented on src/installer/tests/Assets/TestProjects/SingleFileApiTests/Program.cs:
```diff
@@ -1,44 +1,71 @@
-                    Console.WriteLine("APP_CONTEXT_DEPS_FILES: " + deps_files);
-                    return;
+                string arg = args[i];
+                switch (arg)
```
Good point - I added it - and reworked the tests to validate more things.

vitek-karas commented:
There's still a bug in this - the APP_CONTEXT_DEPS_FILES reported with 3.0 backward compat mode points to the app.deps.json as if it was next to the .exe - not into the extraction directory.

jeffschwMSFT commented:
Adding a few additional code reviewers as we are really late in .NET 5.

elinor-fung commented:
The other thing I found is that if there is no .deps.json file (e.g. `GenerateDependencyFile=false`), the published single-file app doesn't run - self-contained can't find coreclr on Windows, can't find SPCL on Linux. Runs fine with 3.1 single-file (or `IncludeAllContentForSelfExtract=false`)

vitek-karas commented:
Sad face...
I guess we could make that work with a fix which I think Mateo actually implemented at one point - the piece of code you asked about (why do we add CoreLib to TPA  at the end, always). We could tweak that to add CoreLib to TPA in the 3.0 compat case (it is currently disabled when CoreLib exists in the bundle, because we expect it to be loaded from the bundle - which is technically wrong).

vitek-karas commented:
That said - I would be fine not supporting that case if it came down to it - it's such a niche thing ... I think

elinor-fung commented:
Yeah, I'd also be fine not supporting it - seems like we'd have to special case where we load coreclr from and adding SPCL to the TPA in compat-mode... I think it is pretty niche as well, but we do talk about how loading works in our [default dependency probing](https://docs.microsoft.com/dotnet/core/dependency-loading/default-probing) doc.

vitek-karas committed:
Fix the reported path to .deps.json for back compat mode

The path must point to the extraction folder and not use the "fake" in-bundle path.

vitek-karas commented:
So the last commit should fix the .deps.json problem (had trouble with local setup, somehow everything got corrupted and I had to rebuild the entire repo again). That leaves only the no .deps.json case - I may look at that as well - since we know about it and the fix seems simple enough - we might as well.

vitek-karas committed:
Fix a test bug on Linux

Actually we should be using path separator char in the APP_CONTEXT_DEPS_FILES, but that's something we won't be able to change really (too much chance of breaking something without too much value).

vitek-karas commented:
I tried the missing .deps.json case and it actually asserts in the host... pretty bad. I will probably look into this a little bit.
On a bright side - it works without .deps.json and without backward compat mode - I don't know why we find the runtime in the host, but after that finding everything else is easy because we probe the bundle every time we need instead of enumerating files in the TPA - so as long as it's in the bundle, it doesn't have to be in .deps.json for runtime to find it.

agocke commented:
@jkotas @janvorli @davidwrighton Would one of you have some time to do a last review here?

jkotas commented:
Changes under src/coreclr/src look good to me. I have glanced over the rest and I did not find anything suspect, but I do not understand that part well.

janvorli commented on src/coreclr/src/binder/utils.cpp:
```diff
@@ -115,4 +117,139 @@ namespace BINDER_SPACE
+            SString sDll(W(".dll"));
+            SString sExe(W(".exe"));
+
+            if (!dllOnly && (outPath.EndsWithCaseInsensitive(sNiDll) ||
```
Why do we exclude the .ni.dll here? I wonder if that's correct.

elinor-fung replied:
The `dllOnly` is currently only set to `true` when searching for System.Private.CoreLib (in the added fallback logic), so I believe the handling of it was more targeted to avoiding extra work in that scenario of just looking for .dll and not .ni.dll. For longer term / 6.0, I expect it will be refactored away per https://github.com/dotnet/runtime/pull/42435#discussion_r493857718.

agocke commented:
@janvorli Any more comments?

janvorli commented:
@agocke no, I don't. Things look reasonable, but I am not much familiar with the code being updated.

agocke commented:
Understandable, thanks for the input!

@vitek-karas @mateoatr @elinor-fung Ready to merge?

vitek-karas commented:
I read through it once more - looks good to me.

elinor-fung commented:
Looks good to me as well.

mateoatr commented:
:shipit:

agocke commented:
/backport to release/5.0
